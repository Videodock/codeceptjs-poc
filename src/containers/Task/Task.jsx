import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { gql } from "apollo-boost";
import PropTypes from 'prop-types';
import client from '../../graphql';
import ActiveTask from '../../components/ActiveTask';

export default function Task({ campaignIdentifier, taskIdentifier }) {
  const history                               = useHistory();
  const taskAmount                            = 10;
  const [loading, setLoading]                 = useState(false);
  const [task, setTask]                       = useState('');
  const [taskIdentifiers, setTaskIdentifiers] = useState([]);

  // Retrieve task from url (or restart)
  useEffect(() => {
    if(taskIdentifier) {
      client.query({ query: CLIENT_NEXT_POTENTIAL_ACTION_QUERY, variables: { taskIdentifiers: [taskIdentifier], offset: 0 } })
        .then(response => {
          const currentTask = response.data.ControlAction[0];

          if(currentTask) {
            setTask(currentTask);
            console.log('set task from url');
          } else {
            history.replace(`/campaign/${campaignIdentifier}/task`);
          }
        });
    }
  }, [campaignIdentifier, history, taskIdentifier]);

  // Retrieve tasks new run
  useEffect(() => {
    if(taskIdentifier || loading) {
      return;
    }

    const getCurrentTask = identifiers => {
      client.query({ query: CLIENT_NEXT_POTENTIAL_ACTION_QUERY, variables: { taskIdentifiers: identifiers, offset: 0 } })
        .then(response => {
          const firstTask = response.data.ControlAction[0];

          console.log('set task from current task');
          if(firstTask) {
            history.replace(`/campaign/${campaignIdentifier}/task/${firstTask.identifier}`);
            setLoading(false);
          }
        });
    };

    setLoading(true);

    client.query({ query: ALL_POTENTIAL_ACTIONS_QUERY, variables: { identifier: campaignIdentifier } })
      .then(response => {
        const allPotentialActions = response.data.ControlAction;
        const offset              = Math.floor(Math.random() * allPotentialActions.length);
        const identifiers         = allPotentialActions.slice(offset, taskAmount).map(({ identifier }) => identifier);

        setTaskIdentifiers(identifiers);
        getCurrentTask(identifiers);
      });
  }, [campaignIdentifier, history, loading, taskIdentifier, taskIdentifiers]);

  if (loading) {
    return <p>Loading ...</p>;
  }

  return <ActiveTask name={task.name} url={task.url} campaignIdentifier={campaignIdentifier} />;
}

Task.propTypes = {
  campaignIdentifier: PropTypes.string,
  taskIdentifier    : PropTypes.string,
};

export const ALL_POTENTIAL_ACTIONS_QUERY = gql`
query AllPotentialActions($identifier: ID!) {
  ControlAction(
    filter: {
      target: { identifier: $identifier }
      actionStatus: PotentialActionStatus
    }
  ) {
    identifier
  }
}`;

export const CLIENT_NEXT_POTENTIAL_ACTION_QUERY = gql`
query ClientNextPotentialAction($taskIdentifiers: [ID!]!, $offset: Int!) {
  ControlAction(
    filter: {
      identifier_in: $taskIdentifiers
      actionStatus: PotentialActionStatus
    }
    first: 1
    offset: $offset
  ) {
    identifier
    name
    url
  }
}`;
