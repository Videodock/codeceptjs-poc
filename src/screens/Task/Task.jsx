import React from 'react';
import { useTranslation } from 'react-i18next';
import { Helmet } from 'react-helmet';
import TaskContainer from '../../containers/Task';

export default function Task ({ match }) {
  const { t }                                  = useTranslation('task');
  const { campaignIdentifier, taskIdentifier } = match.params;

  return (
    <React.Fragment>
      <Helmet>
        <title>{t('task')}</title>
        <meta name="description" content={t('description')} />
        <meta property="og:title" content="Task" />
      </Helmet>
      <TaskContainer campaignIdentifier={campaignIdentifier} taskIdentifier={taskIdentifier} />
    </React.Fragment>
  );
}
