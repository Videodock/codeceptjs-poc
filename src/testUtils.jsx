import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@material-ui/styles';
import { I18nextProvider } from 'react-i18next';
import { ApolloProvider } from '@apollo/client';
import { BrowserRouter as Router } from 'react-router-dom';
import 'url-search-params-polyfill';
import i18n from '../src/i18n';
import client from './graphql';
import theme from './theme';
import store from './redux';

const wrapper = ({ children }) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <I18nextProvider i18n={i18n}>
          <Router>
            {children}
          </Router>
        </I18nextProvider>
      </ThemeProvider>
    </Provider>
  </ApolloProvider>
);

const customRender = (ui, options) => render(ui, { wrapper, ...options });

export { customRender as render };
