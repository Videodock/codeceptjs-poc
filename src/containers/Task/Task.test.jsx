import React from 'react';
import { MockedProvider } from '@apollo/react-testing';
import { wait } from '@testing-library/react';
import { fireEvent } from '@testing-library/react';
import { render } from '../../testUtils';
import Task, { ALL_POTENTIAL_ACTIONS_QUERY, CLIENT_NEXT_POTENTIAL_ACTION_QUERY } from './Task';

const mocks = [
  {
    request: {
      query    : ALL_POTENTIAL_ACTIONS_QUERY,
      variables: {
        identifier: 'e63fc3c5-f84e-4a64-9d5b-98a49dd4680c',
      },
    },
    result: {
      data: {
        ControlAction: [
          {
            identifier: "9754f772-988e-4758-b5a6-f481c0266002",
            name      : "Campaign Manager Task #4",
            url       : "https://demo.videodock.com/trompa/ce-task",
            __typename: "ControlAction",
          },
          {
            identifier: "bbd650f8-2c8d-43a0-ad92-34f8d14f4582",
            name      : "Campaign Manager Task #6",
            url       : "https://demo.videodock.com/trompa/ce-task",
            __typename: "ControlAction",
          },
          {
            identifier: "f67a7ab1-13ae-4d27-a6aa-5f49e0240858",
            name      : "Campaign Manager Task #7",
            url       : "https://demo.videodock.com/trompa/ce-task",
            __typename: "ControlAction",
          },
        ],
      },
    },
  },
  {
    request: {
      query    : CLIENT_NEXT_POTENTIAL_ACTION_QUERY,
      variables: {
        identifier: 'e63fc3c5-f84e-4a64-9d5b-98a49dd4680c',
      },
    },
    result: {
      data: {
        ControlAction: [
          {
            identifier: "f67a7ab1-13ae-4d27-a6aa-5f49e0240858",
            name      : "Campaign Manager Task #7",
            url       : "https://demo.videodock.com/trompa/ce-task",
            __typename: "ControlAction",
          },
        ],
      },
    },
  },
];

describe('<Task />', () => {
  test('matches snapshot', () => {
    const { container } = render((
      <MockedProvider mocks={mocks} addTypename={false}>
        <Task campaignIdentifier="e63fc3c5-f84e-4a64-9d5b-98a49dd4680c" />
      </MockedProvider>));

    expect(container).toMatchSnapshot();
  });

  test('renders loading state initially', () => {
    const { getByText } = render((
      <MockedProvider mocks={[]} addTypename={false}>
        <Task campaignIdentifier="e63fc3c5-f84e-4a64-9d5b-98a49dd4680c" />
      </MockedProvider>));

    wait(() => expect(getByText('Loading...')).toBeTruthy());
  });

  test('renders task without task identifier', () => {
    const { getByText } = render((
      <MockedProvider mocks={mocks} addTypename={false}>
        <Task campaignIdentifier="e63fc3c5-f84e-4a64-9d5b-98a49dd4680c" />
      </MockedProvider>));

    wait(() => expect(getByText('CE Task Page')).toBeTruthy());
  });

  test('renders task with task identifier', () => {
    const { getByText } = render((
      <MockedProvider mocks={mocks} addTypename={false}>
        <Task campaignIdentifier="e63fc3c5-f84e-4a64-9d5b-98a49dd4680c" taskIdentifier="f67a7ab1-13ae-4d27-a6aa-5f49e0240858" />
      </MockedProvider>));

    wait(() => expect(getByText('CE Task Page')).toBeTruthy());
  });
});

