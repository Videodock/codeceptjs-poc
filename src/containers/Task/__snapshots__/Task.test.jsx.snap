// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Task /> matches snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-dense makeStyles-toolbarTop-2 MuiToolbar-gutters"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          class="makeStyles-logo-3"
          src="logo.png"
        />
      </a>
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        href="/campaign/e63fc3c5-f84e-4a64-9d5b-98a49dd4680c"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          Close
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
  </header>
  <iframe
    class="makeStyles-iframe-1"
    name="task-iframe"
    title="Task"
  />
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed makeStyles-appbarBottom-86 MuiAppBar-colorPrimary mui-fixed MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-dense makeStyles-toolbarBottom-87 MuiToolbar-gutters"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Loading...
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;
